pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TopicTransactionBufferTest.java

```diff
@@ -405,4 +412,168 @@ private void assertGetLastMessageId(Consumer<?> consumer, MessageIdImpl expected
+        admin.topics().unload(topic);
+        PersistentTopic persistentTopic2 = (PersistentTopic) pulsarServiceList.get(0).getBrokerService()
+                .getTopic(topic, false)
+                .get()
+                .get();
+        TopicTransactionBuffer topicTransactionBuffer2 = (TopicTransactionBuffer) persistentTopic2
+                .getTransactionBuffer();
+        Awaitility.await().untilAsserted(() -> {
+            Assert.assertEquals(topicTransactionBuffer2.getState(), TopicTransactionBufferState.State.Ready);
+        });
+        consumer.redeliverUnacknowledgedMessages();
+        for (int i = 0; i < 2; i++) {
+            Message<String> message = consumer.receive(5, TimeUnit.SECONDS);
+            Assert.assertEquals(message.getValue(), normalMsg);
+        }
+        Message<String> message = consumer.receive(5, TimeUnit.SECONDS);
+        Assert.assertNull(message);
+    }
+
+    /**
+     * Send some messages before transaction buffer ready and then send some messages after transaction buffer ready,
+     * these messages should be received in order.
+     */
+    @Test
+    public void testMessagePublishInOrder() throws Exception {
+        // 1. Prepare test environment.
+        String topic = "persistent://" + NAMESPACE1 + "/testMessagePublishInOrder" + RandomUtils.nextLong();
+        admin.topics().createNonPartitionedTopic(topic);
+        PersistentTopic persistentTopic = (PersistentTopic) pulsarServiceList.get(0).getBrokerService()
+                .getTopic(topic, false)
+                .get()
+                .get();
+        @Cleanup
+        Producer<Integer> producer = pulsarClient.newProducer(Schema.INT32)
+                .topic(topic)
+                .create();
+        @Cleanup
+        Consumer<Integer> consumer = pulsarClient.newConsumer(Schema.INT32)
+                .topic(topic)
+                .subscriptionName("sub")
+                .subscribe();
+        Transaction transaction = pulsarClient.newTransaction()
+                .withTransactionTimeout(5, TimeUnit.HOURS)
+                .build().get();
+        // 2. Set a new future in transaction buffer as `transactionBufferFuture` to stimulate whether the
+        // transaction buffer recover completely.
+        TopicTransactionBuffer topicTransactionBuffer = (TopicTransactionBuffer) persistentTopic.getTransactionBuffer();
+        CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+        Whitebox.setInternalState(topicTransactionBuffer, "transactionBufferFuture", completableFuture);
+        Field stateField = TopicTransactionBufferState.class.getDeclaredField("state");
+        stateField.setAccessible(true);
+        stateField.set(topicTransactionBuffer, TopicTransactionBufferState.State.Ready);
+        // Register this topic to the transaction in advance to avoid the sending request pending here.
+        ((TransactionImpl) transaction).registerProducedTopic(topic).get(5, TimeUnit.SECONDS);
```

Comment:
It's not good to just introduce one more dependency for just one API. I also don't recommend modifying a final field via reflection. It would be better to mock methods rather than fields.

See https://stackoverflow.com/questions/23629766/how-to-mock-final-field-mockito-powermock

A proper way could be implementing a customized `TransactionBuffer` and call `PulsarService#setTransactionBufferProvider` to modify the TB provider. In this case, you can modify the internal stats like `transactionBufferFuture` and `state` without any reflection.

Reply:
Indeed, you are right. Although reflection is simple to use, it can achieve the purpose of our testing with as little code as possible. But this will introduce additional dependencies, even if the scope of this dependency is test.
Reflection should not be our first choice. I will find time to re-implement a `TransactionBuffer` for this test.